### PR TITLE
c++ api for keyword spotter

### PR DESCRIPTION
**Pull request by xinhecuican**

### CODE REVIEW

File: sherpa-onnx/c-api/c-api.cc [hunk 1]
```diff
@@ -410,6 +411,184 @@ void DestroyOfflineRecognizerResult(
...
+  spotter_config.keywords_threshold =
+      SHERPA_ONNX_OR(config->keywords_threshold, 0.25);
+
+  spotter_config.keywords_file =
+      SHERPA_ONNX_OR(config->keywords_file, "");
+
+  if (config->model_config.debug) {
+    SHERPA_ONNX_LOGE("%s\n", spotter_config.ToString().c_str());
+  }
+
```
Comment (csukuangfj) [inline sherpa-onnx/c-api/c-api.cc:475]: Please add
```cpp
  if (!spotter_config.Validate()) {
    SHERPA_ONNX_LOGE("Errors in config!");
    return nullptr;
  }
```

File: sherpa-onnx/c-api/c-api.cc [hunk 2]
```diff
@@ -410,6 +411,184 @@ void DestroyOfflineRecognizerResult(
...
+void DecodeKeywordStream(SherpaOnnxKeywordSpotter* spotter,
+                         SherpaOnnxOnlineStream* stream) {
+  return spotter->impl->DecodeStream(stream->impl.get());
+}
+
+void DecodeMultipleKeywordStreams(
+    SherpaOnnxKeywordSpotter *spotter, SherpaOnnxOnlineStream **streams,
+    int32_t n) {
+  std::vector<sherpa_onnx::OnlineStream*> ss(n);
+  for(int32_t i=0; i!=n; ++i) {
```
Comment (csukuangfj) [inline sherpa-onnx/c-api/c-api.cc]:
```suggestion
  for(int32_t i = 0; i != n; ++i) {
```

Please use `clang-format` to format your code.


You can install `clang-format` with
```
pip install clang-format
```

File: sherpa-onnx/c-api/c-api.h [hunk 3]
```diff
@@ -473,6 +473,123 @@ SHERPA_ONNX_API const SherpaOnnxOfflineRecognizerResult *GetOfflineStreamResult(
...
+  /// For instance, for BPE-based models it consists of a list of BPE tokens.
+  const char* tokens;
+
+  const char* const* tokens_arr;
+
+  int32_t count;
+
+  /// timestamps.size() == tokens.size()
+  /// timestamps[i] records the time in seconds when tokens[i] is decoded.
+  float* timestamps;
```
Comment (csukuangfj) [inline sherpa-onnx/c-api/c-api.h:497]:
```suggestion
  const float* timestamps;
```

File: sherpa-onnx/c-api/c-api.h [hunk 4]
```diff
@@ -473,6 +473,123 @@ SHERPA_ONNX_API const SherpaOnnxOfflineRecognizerResult *GetOfflineStreamResult(
...
+   * The returned string contains:
+   *   {
+   *     "keyword": "The triggered keyword",
+   *     "tokens": [x, x, x],
+   *     "timestamps": [x, x, x],
+   *     "start_time": x,
+   *   }
+   */
+  const char* json;
+}SherpaOnnxKeywordResult ;
```
Comment (csukuangfj) [inline sherpa-onnx/c-api/c-api.h]:
```suggestion
} SherpaOnnxKeywordResult ;
```

**Review by csukuangfj — COMMENTED** [review]
Thanks! Left some minor comments.

File: sherpa-onnx/c-api/c-api.h [hunk 5]
```diff
@@ -473,6 +473,123 @@ SHERPA_ONNX_API const SherpaOnnxOfflineRecognizerResult *GetOfflineStreamResult(
...
+/// @param p A pointer returned by CreateKeywordSpotter()
+SHERPA_ONNX_API void DestroyKeywordSpotter(
+    SherpaOnnxKeywordSpotter* spotter);
+
+/// Create an online stream for accepting wave samples.
+///
+/// @param spotter A pointer returned by CreateKeywordSpotter()
+/// @return Return a pointer to an OnlineStream. The user has to invoke
+///         DestroyOnlineStream() to free it to avoid memory leak.
+SHERPA_ONNX_API SherpaOnnxOnlineStream* CreateKeywordStream(
```
Comment (csukuangfj) [inline sherpa-onnx/c-api/c-api.h:546]: Could you prefix every function name with `SherpaOnnx`?
That is, replace `CreateKeywordStream` with `SherpaOnnxCreateKeywordStream`.

---

I think it was an error to not use `SherpaOnnx` for online and offline ASR C APIs.

**Comment by csukuangfj** [comment]
Please use the following command to check the code style locally
```bash
cd sherpa-onnx
./scripts/check_style_cpplint.sh 2
```

File: sherpa-onnx/c-api/c-api.h (hunk 5 above)
Comment (xinhecuican) [reply sherpa-onnx/c-api/c-api.h:546]: after add this many line exceed 80 char, I tried the style is not uniform
and change api name means must modify code of other project

**Comment by lovemefan** [comment]
> This PR has conflicts with #648 since both of them add C API for KWS.
> 
> @lovemefan What do you think about this PR?

ok, I'm going to copy the code of c api from this PR

**Comment by csukuangfj** [comment]
> > This PR has conflicts with #648 since both of them add C API for KWS.
> > @lovemefan What do you think about this PR?
> 
> ok, I'm going to copy the code of c api from this PR

Thanks! I am merging this PR. Please merge the latest master to your branch after we merge this PR.

**Comment by csukuangfj** [comment]
@xinhecuican Thank you for your first-time contribution!

**Pull request merged**